### ISSUE TITLE
Compose V2: Cannot run docker-compose if a container uses `build:`

### ISSUE THREAD

**Issue by WhyNotHugo** [containers/podman#11874]
**Is this a BUG REPORT or FEATURE REQUEST? (leave only one on its own line)**

/kind bug

**Description**

If any container has a `build:` directive, `docker-compose` fails to start.

**Steps to reproduce the issue:**

```dockerfile
FROM alpine:latest

RUN echo 1
```

```yaml
services:
  test:
    build: .
    command: ls /app
```

**Describe the results you received:**

```console
$ docker-compose up
[+] Building 0.0s (0/0)
failed to get status: rpc error: code = Unavailable desc = connection error: desc = "transport: Error while dialing unable to upgrade to h2c, received 404"
```

**Describe the results you expected:**

This should build and run.

**Additional information you deem important (e.g. issue happens only occasionally):**

Works fine with docker-compose.

**Output of `podman version`:**

```console
$ podman --version
dpodman version 3.4.0
$ docker-compose --version
Docker Compose version 2.0.1
```

**Output of `podman info --debug`:**

```
host:
  arch: amd64
  buildahVersion: 1.23.1
  cgroupControllers: []
  cgroupManager: systemd
  cgroupVersion: v2
  conmon:
    package: /usr/bin/conmon is owned by conmon 1:2.0.30-1
    path: /usr/bin/conmon
    version: 'conmon version 2.0.30, commit: 2792c16f4436f1887a7070d9ad99d9c29742f38a'
  cpus: 8
  distribution:
    distribution: arch
    version: unknown
  eventLogger: journald
  hostname: victory
  idMappings:
    gidmap:
    - container_id: 0
      host_id: 1000
      size: 1
    - container_id: 1
      host_id: 100000
      size: 65537
    uidmap:
    - container_id: 0
      host_id: 1000
      size: 1
    - container_id: 1
      host_id: 100000
      size: 65537
  kernel: 5.14.9-arch2-1
  linkmode: dynamic
  logDriver: journald
  memFree: 234930176
  memTotal: 16462659584
  ociRuntime:
    name: crun
    package: /usr/bin/crun is owned by crun 1.1-1
    path: /usr/bin/crun
    version: |-
      crun version 1.1
      commit: 5b341a145c4f515f96f55e3e7760d1c79ec3cf1f
      spec: 1.0.0
      +SYSTEMD +SELINUX +APPARMOR +CAP +SECCOMP +EBPF +YAJL
  os: linux
  remoteSocket:
    exists: true
    path: /run/user/1000/podman/podman.sock
  security:
    apparmorEnabled: false
    capabilities: CAP_CHOWN,CAP_DAC_OVERRIDE,CAP_FOWNER,CAP_FSETID,CAP_KILL,CAP_NET_BIND_SERVICE,CAP_SETFCAP,CAP_SETGID,CAP_SETPCAP,CAP_SETUID,CAP_SYS_CHROOT
    rootless: true
    seccompEnabled: true
    seccompProfilePath: /etc/containers/seccomp.json
    selinuxEnabled: false
  serviceIsRemote: false
  slirp4netns:
    executable: /usr/bin/slirp4netns
    package: /usr/bin/slirp4netns is owned by slirp4netns 1.1.12-1
    version: |-
      slirp4netns version 1.1.12
      commit: 7a104a101aa3278a2152351a082a6df71f57c9a3
      libslirp: 4.6.1
      SLIRP_CONFIG_VERSION_MAX: 3
      libseccomp: 2.5.2
  swapFree: 19307757568
  swapTotal: 34343288832
  uptime: 15h 3m 4.92s (Approximately 0.62 days)
plugins:
  log:
  - k8s-file
  - none
  - journald
  network:
  - bridge
  - macvlan
  volume:
  - local
registries:
  search:
  - docker.io
store:
  configFile: /home/hugo/.config/containers/storage.conf
  containerStore:
    number: 18
    paused: 0
    running: 0
    stopped: 18
  graphDriverName: overlay
  graphOptions: {}
  graphRoot: /home/hugo/.local/share/containers/storage
  graphStatus:
    Backing Filesystem: extfs
    Native Overlay Diff: "true"
    Supports d_type: "true"
    Using metacopy: "false"
  imageStore:
    number: 30
  runRoot: /run/user/1000/containers
  volumePath: /home/hugo/.local/share/containers/storage/volumes
version:
  APIVersion: 3.4.0
  Built: 1633083258
  BuiltTime: Fri Oct  1 12:14:18 2021
  GitCommit: 6e8de00bb224f9931d7402648f0177e7357ed079
  GoVersion: go1.17.1
  OsArch: linux/amd64
  Version: 3.4.0
```

**Package info (e.g. output of `rpm -q podman` or `apt list podman`):**

```
Podman 3.4.0-1 (Arch package).
```

**Have you tested with the latest version of Podman and have you checked the Podman Troubleshooting Guide? (https://github.com/containers/podman/blob/master/troubleshooting.md)**

Yes

**Additional environment details (AWS, VirtualBox, physical, etc.):**

Physical host, running rootless.

**Comment by Luap99** [comment]
Yes, docker-compose V2 does not work at the moment. I recommend to use the 1.X python based version for now.

**Comment by WhyNotHugo** [comment]
still relevant.

**Comment by brnosouza** [comment]
Still relevant.

**Comment by heapifyman** [comment]
Looks like this has been fixed in the meantime?

I don't get any errors when running the steps to reproduce:

```
docker-compose up                                                                                     
Sending build context to Docker daemon     221B
STEP 1/2: FROM alpine:latest
STEP 2/2: RUN echo 1
--> Using cache 1c551f141a7f83bb58b76fcd5f7295ae5742c1a5f66ebfa7352ba961581c35b0
COMMIT podman-test_test
--> 1c551f141a7
Successfully tagged localhost/podman-test_test:latest
Successfully tagged localhost/downloads_test:latest
1c551f141a7f83bb58b76fcd5f7295ae5742c1a5f66ebfa7352ba961581c35b0
Successfully built 1c551f141a7f
Successfully tagged podman-test_test
[+] Running 2/0
 ⠿ Network podman-test_default   Created                                                                                               0.0s
 ⠿ Container podman-test-test-1  Created                                                                                               0.0s
Attaching to podman-test-test-1
podman-test-test-1  | ls: /app: No such file or directory
```

podman version:

```
Version:      3.4.4
API Version:  3.4.4
Go Version:   go1.17.4
Git Commit:   f6526ada1025c2e3f88745ba83b8b461ca659933
Built:        Thu Dec  9 19:30:40 2021
OS/Arch:      linux/amd64
```

docker-compose version:

```
Docker Compose version 2.2.3
```

**Comment by brnosouza** [comment]
@heapifyman I tested with both podman 3.4.4 and  4.0.0 and it is not working with compose v2, did you change something?

**Comment by heapifyman** [comment]
@brnosouza I don't think so. I just copy pasted the Dockerfile and docker-compose.yml content from @WhyNotHugo and ran `docker-compose up`.

I am using Manjaro Linux. I have installed packages:

* podman
* podman-dnsname
* podman-docker
* docker-compose

from official Manjaro / Arch community repository.

In addition, I have the following in `~/.config/containers/registries.conf̀`:

```
unqualified-search-registries = ["docker.io"]
```

And the following environment variable - as described at https://wiki.archlinux.org/title/Podman#Docker_Compose:

```
DOCKER_HOST=unix:///run/user/1000/podman/podman.sock
```

Output of `podman info --debug̀` looks pretty much the same as the one posted by ̀@WhyNotHugo. Version numbers are of course different and I am on kernel `5.16.7-1-MANJARO`.

**Comment by brnosouza** [comment]
@heapifyman I tested using docker-compose v2.2.3 and podman 4, and the output was kinda the same:

```shell
$ docker-compose up

failed to get status: rpc error: code = Unavailable desc = connection error: desc = "transport: Error while dialing unable to upgrade to h2c, received 404"
```

I'm using Fedora 35 in this case 🤔

**Comment by mheon** [comment]
We've fixed the other notable Compose v2 issue, and added Compose v2.2.3 to our test suite (which includes several compose projects with builds), and it seems to be working fine so far. Given this, I'm going to close this, but if anyone is able to reproduce with Podman's main branch and Compose v2.2.3, I'd be very interested in seeing a reproducer.

**Comment by CRThaze** [comment]
For anyone else encountering this issue:
The resolution seems to currently be downgrading to `docker-compose` `v2.2.3`

On macOS the downgrade can be performed like this:
```bash
brew uninstall docker-compose
wget https://raw.githubusercontent.com/Homebrew/homebrew-core/bf19d2715bbab845a440d411aa9281c776fa2fcf/Formula/docker-compose.rb
brew install docker-compose.rb
rm docker-compose.rb
brew pin docker-compose
```

As of the time of this writing the `"transport: Error while dialing unable to upgrade to h2c, received 404"` seems to affect most versions after `v2.2.3` up to and including `v2.4.1`

**Comment by mheon** [comment]
Can you open a fresh issue? I doubt that `docker compose` reverted their legacy build support. Possible they now default to Buildkit and it needs to be forced to use the standard build API? I can look into it.

**Comment by nyonson** [comment]
Fresh issue: https://github.com/containers/podman/issues/13889

**Comment by sanchouz12** [comment]
> For anyone else encountering this issue: The resolution seems to currently be downgrading to `docker-compose` `v2.2.3`
> 
> On macOS the downgrade can be performed like this:
> 
> ```shell
> brew uninstall docker-compose
> wget https://raw.githubusercontent.com/Homebrew/homebrew-core/bf19d2715bbab845a440d411aa9281c776fa2fcf/Formula/docker-compose.rb
> brew install docker-compose.rb
> rm docker-compose.rb
> brew pin docker-compose
> ```
> 
> As of the time of this writing the `"transport: Error while dialing unable to upgrade to h2c, received 404"` seems to affect most versions after `v2.2.3` up to and including `v2.4.1`

I used latest docker-compose version, currently it's `2.6.1`, on CentOS 7 production server and got this issue. Solved by downgrading straight to `2.2.2`
